src/oscore/oscore_context.c

```diff
@@ -0,0 +1,787 @@
+  osc_ctx->save_seq_num_func_param = o_osc_ctx->save_seq_num_func_param;
+
+  if (o_osc_ctx->master_secret) {
+    /* sender_ key */
+    sender_ctx->sender_key = oscore_build_key(osc_ctx,
+                                              sender_id,
+                                              coap_make_str_const("Key"),
+                                              CONTEXT_KEY_LEN);
+    if (!sender_ctx->sender_key)
+      goto error;
+
+    /* common IV */
+    osc_ctx->common_iv = oscore_build_key(osc_ctx,
+                                          NULL,
+                                          coap_make_str_const("IV"),
+                                          CONTEXT_INIT_VECT_LEN);
+    if (!osc_ctx->common_iv)
+      goto error;
+  }
+
+  /*
+   * Need to set the last Sender Seq Num based on ssn_freq
+   * The value should only change if there is a change to ssn_freq
+   * and (potentially) be lower than seq, then save_seq_num_func() is
+   * immediately called on next SSN update.
+   */
+  sender_ctx->next_seq = 0;
+  sender_ctx->seq = 0;
+
+  sender_ctx->sender_id = coap_new_bin_const(sender_id->s, sender_id->length);
+
+  copy_rid = coap_new_bin_const(recipient_id->s, recipient_id->length);
+  if (copy_rid == NULL)
+    goto error;
+  if (oscore_add_recipient(osc_ctx, copy_rid, 0) == NULL)
+    goto error;
+
+  oscore_log_context(osc_ctx, "New Common context");
+  oscore_enter_context(c_context, osc_ctx);
+
+  return osc_ctx;
+
+error:
+  oscore_free_context(osc_ctx);
+  return NULL;
+}
+
+oscore_ctx_t *
+oscore_derive_ctx(coap_context_t *c_context, coap_oscore_conf_t *oscore_conf) {
+  oscore_ctx_t *osc_ctx = NULL;
+  oscore_sender_ctx_t *sender_ctx = NULL;
+  size_t i;
+
+  osc_ctx = coap_malloc_type(COAP_OSCORE_COM, sizeof(oscore_ctx_t));
+  if (osc_ctx == NULL)
+    goto error;
+  memset(osc_ctx, 0, sizeof(oscore_ctx_t));
+
+  sender_ctx = coap_malloc_type(COAP_OSCORE_SEN, sizeof(oscore_sender_ctx_t));
+  if (sender_ctx == NULL)
+    goto error;
+  memset(sender_ctx, 0, sizeof(oscore_sender_ctx_t));
+
+  osc_ctx->sender_context = sender_ctx;
+  osc_ctx->master_secret = oscore_conf->master_secret;
+  osc_ctx->master_salt = oscore_conf->master_salt;
+  osc_ctx->aead_alg = oscore_conf->aead_alg;
+  osc_ctx->hkdf_alg = oscore_conf->hkdf_alg;
+  osc_ctx->id_context = oscore_conf->id_context;
+  osc_ctx->ssn_freq = oscore_conf->ssn_freq ? oscore_conf->ssn_freq : 1;
+  osc_ctx->replay_window_size = oscore_conf->replay_window
+                                    ? oscore_conf->replay_window
+                                    : COAP_OSCORE_DEFAULT_REPLAY_WINDOW;
+  osc_ctx->rfc8613_b_1_2 = oscore_conf->rfc8613_b_1_2;
+  osc_ctx->rfc8613_b_2 = oscore_conf->rfc8613_b_2;
+  osc_ctx->save_seq_num_func = oscore_conf->save_seq_num_func;
+  osc_ctx->save_seq_num_func_param = oscore_conf->save_seq_num_func_param;
+
+  if (oscore_conf->master_secret) {
+    /* sender_ key */
+    if (oscore_conf->break_sender_key)
+      /* Interop testing */
```

Comment:
Do we need a configuration option for interop test support?

Reply:
Not sure at what level you mean here - at the app level (well, for the client talking to the interop server), an additional option in the configuration file as a guard against inadvertently setting break_sender_key (or break_recipient_key) or what?

I eventually decided that it was right to document these options in man/coap-oscore-conf.txt.in.